### PR TITLE
Use unicode for cmdline printing

### PR DESCRIPTION
**Pull request by keszybz**
This was much more work and much more churn than I expected. It makes `systemctl status`, `systemc-cgls`, …, treat process commandlines as unicode and process them as such. We don't strip multi-byte characters anymore, and use unicode replacement characters and ellipses. Partial compatibility with non-utf8 locales is preserved, but it only makes limited sense (because characters in one locale and not printable in another locale, so there's no way to make "just work" like with unicode and always something will be randomly broken).

### CODE REVIEW

**Comment by poettering** [comment]
merge conflicts...

**Comment by poettering** [comment]
also, fails all CI

**Comment by keszybz** [comment]
Yep, there was a memleak. Should be fixed now. Let's see if the CI passes.

**Comment by keszybz** [comment]
I now see that the TODO entry is partially implemented (SYSTEMD_LOG_LOCATION= exists). I'll remove it in the next update.

File: src/basic/utf8.c [hunk 1]
```diff
@@ -196,42 +196,86 @@ char *utf8_escape_invalid(const char *str) {
         return p;
 }
 
-char *utf8_escape_non_printable(const char *str) {
-        char *p, *s;
+static int utf8_char_console_width(const char *str) {
+        char32_t c;
+
+        if (utf8_encoded_to_unichar(str, &c) < 0)
+                return -1;
```
Comment (poettering) [inline src/basic/utf8.c]: why not propagate the original error here (i.e. -EINVAL)?

File: src/basic/process-util.c [hunk 2]
```diff
@@ -46,6 +46,11 @@
 #include "user-util.h"
 #include "utf8.h"
 
+/* The kernel limits userspace processes to TASK_COMM_LEN (16 bytes), but allows higher values for its own
+ * workers, e.g. "kworker/u9:3-kcryptd/253:0". Let's pick a fixed smallish limit that will work for the kernel.
```
Comment (poettering) [inline src/basic/process-util.c:51]: oh, jeezuz fuckin christ

File: src/basic/process-util.c [hunk 3]
```diff
@@ -100,22 +101,24 @@ int get_process_comm(pid_t pid, char **ret) {
         return 0;
 }
 
-int get_process_cmdline(pid_t pid, size_t max_length, bool comm_fallback, char **line) {
+int get_process_cmdline(pid_t pid, unsigned max_columns, bool comm_fallback, char **line) {
```
Comment (poettering) [inline src/basic/process-util.c]: why `unsigned` here? size_t sounds better to me still

File: src/basic/escape.c [hunk 4]
```diff
@@ -368,33 +368,71 @@ int cunescape(const char *s, UnescapeFlags flags, char **ret) {
...
+                }
+
+                /* We might need to go back two cycles to fit three dots, so remember two positions */
+                prev2 = prev;
+                prev = tmp_t;
         }
 
-        *t = 0;
+        /* We can just write where we want, since chars are one-byte */
+        size_t c = MIN(console_width, 3u); /* If the console is too narrow, write less dots */
```
Comment (poettering) [inline src/basic/escape.c]: less → fewer?

File: src/basic/utf8.c (hunk 1 above)
Comment (keszybz) [reply src/basic/utf8.c]: I just moved that code around. But I agree, propagating is better. Will fix.

**Comment by keszybz** [comment]
Updated with the all the requested changes.
One more commit one top: call `str_realloc()` in utf8.c.

**Pull request merged**